File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans = []
        path = []
        curSum = 0
        def dfs(root):
            if root is None:
                return
            
            path.append(root.val)
            nonlocal curSum
            curSum += root.val

            if root.left is None and root.right is None:
                if curSum == targetSum:
                    ans.append(path.copy())
            else:
                dfs(root.left)
                dfs(root.right)
            
            path.pop()
            curSum -= root.val
        
        dfs(root)
        return ans

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> res;
        if(!root) return res;
        vector<int> tmp;
        auto dfs = [&](auto &self, TreeNode* node, int curSum) -> void {
            if(curSum == targetSum && !node->left && !node->right) res.push_back(tmp);
            if(node->left) {
                tmp.push_back(node->left->val);
                self(self, node->left, curSum + node->left->val);
                tmp.pop_back();
            }
            if(node->right) {
                tmp.push_back(node->right->val);
                self(self, node->right, curSum + node->right->val);
                tmp.pop_back();
            }
        };
        tmp.push_back(root->val);
        dfs(dfs, root, root->val);
        return res;
    }
};